src/main/java/fi/aalto/cs/apluscourses/presentation/filter/LogicalFilter.java

```diff
@@ -0,0 +1,32 @@
+package fi.aalto.cs.apluscourses.presentation.filter;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BinaryOperator;
+import org.jetbrains.annotations.NotNull;
+
+public class LogicalFilter implements Filter {
+  private final List<? extends Filter> filters;
+  private final BinaryOperator<Boolean> operator;
+
+  public LogicalFilter(List<? extends Filter> filters, BinaryOperator<Boolean> operator) {
+    this.filters = filters;
+    this.operator = operator;
+  }
+
+  @Override
+  @NotNull
+  public Optional<Boolean> apply(Object item) {
+    return filters.stream()
+        .map(filter -> filter.apply(item))
+        .filter(Optional::isPresent)
+        .map(Optional::get)
+        .reduce(operator);
```

Comment:
abstract stuff is a bit hard to read, but guess there's no other way, or... maybe a comment? :open_mouth: